### PR TITLE
[SYCL][CUDA] Add basic sub-group functionality

### PR DESCRIPTION
**Pull request by Pennycook**
Implements:
- Sub-group local id
- Sub-group id
- Number of sub-groups
- Sub-group size
- Max sub-group size

The implementations are functionally correct, but may benefit from
additional optimization.

Signed-off-by: John Pennycook <john.pennycook@intel.com>

---

The implementation is different to the one proposed in https://intel.github.io/llvm-docs/cuda/opencl-subgroup-vs-cuda-crosslane-op.html, because I don't think `sreg.warpid` and `sreg.nwarpid` have the correct semantics for sub-groups.  The mapping from work-items to sub-groups is invariant during a kernel's execution, which isn't true of the warp ID in PTX.  As far as I can tell, the number of warp IDs represents the maximum number of warps that can execute in a CTA rather than the number of warps in a CTA.

NVIDIA's [PTX documentation](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#special-registers-warpid) says that `tid` should be used to compute a "virtual warp ID" if one is required, which is what I've implemented.  I convinced myself that we have to compute the sub-group IDs and sizes from the linear size of the work-group, and couldn't find a simpler way to express this.

Ideally, we wouldn't have to re-compute each of these values on every call.  It would be sufficient to compute them once at the start of the kernel and then re-use them, but I don't have enough knowledge of Clang/LLVM/libclc to implement that.

### CODE REVIEW

**Comment by bader** [comment]
Tagging @Naghasan.

File: libclc/ptx-nvidiacl/libspirv/workitem/get_sub_group_size.cl [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+#include <spirv/spirv.h>
+_CLC_DEF _CLC_OVERLOAD uint __spirv_SubgroupId();
+_CLC_DEF _CLC_OVERLOAD uint __spirv_NumSubgroups();
+_CLC_DEF _CLC_OVERLOAD uint __spirv_SubgroupMaxSize();
+
+_CLC_DEF _CLC_OVERLOAD uint __spirv_SubgroupSize() {
+  if (__spirv_SubgroupId() != __spirv_NumSubgroups() - 1) {
+    return __spirv_SubgroupMaxSize();
+  } else {
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/workitem/get_sub_group_size.cl:17]:
```suggestion
  }
```
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

File: libclc/ptx-nvidiacl/libspirv/workitem/get_sub_group_size.cl [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <spirv/spirv.h>
+_CLC_DEF _CLC_OVERLOAD uint __spirv_SubgroupId();
+_CLC_DEF _CLC_OVERLOAD uint __spirv_NumSubgroups();
+_CLC_DEF _CLC_OVERLOAD uint __spirv_SubgroupMaxSize();
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/workitem/get_sub_group_size.cl:12]: Probably, we should move these forward declarations to corresponding header file similar to https://github.com/intel/llvm/tree/sycl/libclc/generic/include/spirv/workitem to following existing style.

File: libclc/ptx-nvidiacl/libspirv/workitem/get_sub_group_size.cl (hunk 1 above)
Comment (Pennycook) [reply libclc/ptx-nvidiacl/libspirv/workitem/get_sub_group_size.cl]: Fixed in 41d157de0f0f57c850ecc5a54c57de338a87d9a4.

File: libclc/ptx-nvidiacl/libspirv/workitem/get_sub_group_size.cl (hunk 2 above)
Comment (Pennycook) [reply libclc/ptx-nvidiacl/libspirv/workitem/get_sub_group_size.cl]: Makes sense.  I assumed this was an auto-generated SPIR-V header coming in from somewhere else.  Fixed in fcc055a9cfe1c8b935eb6d1c6e7a221f4893edff.

**Review by bader — APPROVED** [review]

**Comment by illuhad** [comment]
> NVIDIA's PTX documentation says that tid should be used to compute a "virtual warp ID" if one is required, which is what I've implemented. I convinced myself that we have to compute the sub-group IDs and sizes from the linear size of the work-group, and couldn't find a simpler way to express this.

Yes, I also stumbled across that - that's also how it works in hipSYCL. Additionally, my understanding is that the laneid register may be less performant than computing the subgroup local id yourself (which is what hipSYCL does), see e.g. https://stackoverflow.com/questions/44337309/whats-the-most-efficient-way-to-calculate-the-warp-id-lane-id-in-a-1-d-grid

**Review by sergey-semenov — APPROVED** [review]

**Comment by Pennycook** [comment]
Thanks, @illuhad.

> Additionally, my understanding is that the laneid register may be less performant than computing the subgroup local id yourself (which is what hipSYCL does), see e.g. https://stackoverflow.com/questions/44337309/whats-the-most-efficient-way-to-calculate-the-warp-id-lane-id-in-a-1-d-grid

We should definitely revisit this if benchmarks suggest that manually computing the lane id is always better.  I found some references to special registers potentially having high latency, but nothing about the relative cost of different special registers.  Since we'd need to read the `tid` and `ntid` special registers to compute things manually, I wasn't confident that I understood the trade-off here.

**Pull request merged**